backstory/src/test/java/com/google/sps/story/DatamuseRequestClientTest.java

```diff
@@ -0,0 +1,337 @@
+   * https://claritysoftware.co.uk/mocking-javas-url-with-mockito/.
+   */
+  @Before
+  public void reset() {
+    httpUrlStreamHandler.resetConnections();
+  }
+
+  /** 
+   * Check that IllegalArgumentException thrown for a multiple
+   * word noun.
+   */
+  @Test
+  public void multipleWordNoun() throws Exception {
+    final String MULTIWORD_NOUN = "multiword noun";
+    final int DEFAULT_MAX = 5;
+
+    DatamuseRequestClient client = new DatamuseRequestClient(DEFAULT_URL);
+
+    try {
+      client.fetchRelatedAdjectives(MULTIWORD_NOUN, DEFAULT_MAX);
+    } catch (IllegalArgumentException exception) {
+      return; // test should pass if this exception is thrown
+    } 
+
+    Assert.fail("IllegalArgumentException was not thrown.");
+  }
+
+  /** 
+   * Check that IllegalArgumentException thrown
+   * for a noun with whitespace.
+   */
+  @Test
+  public void nounWithWhitespace() throws Exception {
+    final String WHITESPACE_NOUN = "noun\n";
+    final int DEFAULT_MAX = 5;
+
+    DatamuseRequestClient client = new DatamuseRequestClient(DEFAULT_URL);
+
+    try {
+      client.fetchRelatedAdjectives(WHITESPACE_NOUN, DEFAULT_MAX);
+    } catch (IllegalArgumentException exception) {
+      return; // test should pass if this exception is thrown
+    } 
+
+    Assert.fail("IllegalArgumentException was not thrown.");
+  }
+
+  /** 
+   * Check that APINotAvailableException is thrown 
+   * if there is no URLConnection injected (which is equivalent
+   * to if it can't access the Datamuse API).
+   */
+  @Test 
+  public void apiNotAvailable() throws Exception {
+    final String NOUN = "beach";
+    final int DEFAULT_MAX = 5;
+
+    DatamuseRequestClient client = new DatamuseRequestClient(DEFAULT_URL);
+
+    try {
+      client.fetchRelatedAdjectives(NOUN, DEFAULT_MAX);
+    } catch (APINotAvailableException exception) {
+      return; // test should pass if this exception is thrown
+    }
+
+    Assert.fail("APINotAvailableException was not thrown.");    
+  }
+
+  /** 
+   * Check that fetchRelatedAdjectives() throws a RuntimeException
+   * when the JSON can't be parsed with the correct message signaling
+   * it's an issue with the JSON.
+   */
+  @Test 
+  public void jsonCannotBeParsed() throws Exception {
+    final String NOUN = "beach";
+    final int MAX_ADJECTIVES = 5;
+    final String QUERIES = "rel_jjb=" + NOUN + "&max=" + MAX_ADJECTIVES;
+    final String INPUT = "not JSON";
+
+    DatamuseRequestClient client = new DatamuseRequestClient(DEFAULT_URL);
+
+    // inject the url but with a response that's not proper JSON
+    injectURLConnectionMock(DEFAULT_URL, QUERIES, INPUT);
+
+    try {
+      String[] actualOutput = client.fetchRelatedAdjectives(NOUN, MAX_ADJECTIVES);
+    } catch (RuntimeException exception) {
+      final String EXPECTED_MESSAGE = "Could not parse the JSON received back from the Datamuse Query.";
+    
+      Assert.assertEquals(EXPECTED_MESSAGE, exception.getMessage());
```

Comment:
I'd discourage asserting the exact value of a string. This type of thing is likely to change, and asserting against this could have an undesirable maintenance cost.

Also applies below.

Reply:
Ok that makes sense. I'll remove the error message validation.

Reply:
Done